src/Spanner/ValueMapper.php

```diff
@@ -0,0 +1,328 @@
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Spanner;
+
+use Google\Cloud\Core\ArrayTrait;
+use Google\Cloud\Core\Int64;
+use google\spanner\v1\TypeCode;
+
+/**
+ * Manage value mappings between Google Cloud PHP and Cloud Spanner
+ */
+class ValueMapper
+{
+    use ArrayTrait;
+
+    const NANO_REGEX = '/\.(\d{1,9})Z/';
+
+    const TYPE_BOOL = TypeCode::TYPE_BOOL;
+    const TYPE_INT64 = TypeCode::TYPE_INT64;
+    const TYPE_FLOAT64 = TypeCode::TYPE_FLOAT64;
+    const TYPE_TIMESTAMP = TypeCode::TYPE_TIMESTAMP;
+    const TYPE_DATE = TypeCode::TYPE_DATE;
+    const TYPE_STRING = TypeCode::TYPE_STRING;
+    const TYPE_BYTES = TypeCode::TYPE_BYTES;
+    const TYPE_ARRAY = TypeCode::TYPE_ARRAY;
+    const TYPE_STRUCT = TypeCode::TYPE_STRUCT;
+    const TYPE_NULL = 'null';
+
+    /**
+     * @var bool
+     */
+    private $returnInt64AsObject;
+
+    /**
+     * @param bool $returnInt64AsObject
+     */
+    public function __construct($returnInt64AsObject)
+    {
+        $this->returnInt64AsObject = $returnInt64AsObject;
+    }
+
+    /**
+     * Accepts an array of key/value pairs, where the key is a SQL parameter
+     * name and the value is the value interpolated by the server, and returns
+     * an array of parameters and inferred parameter types.
+     *
+     * @param array $parameters The key/value parameters.
+     * @return array An associative array containing params and paramTypes.
+     */
+    public function formatParamsForExecuteSql(array $parameters)
+    {
+        $paramTypes = [];
+
+        foreach ($parameters as $key => $value) {
+            list ($parameters[$key], $paramTypes[$key]) = $this->paramType($value);
+        }
+
+        return [
+            'params' => $parameters,
+            'paramTypes' => $paramTypes
+        ];
+    }
+
+    /**
+     * Accepts a list of values and encodes the value into a format accepted by
+     * the Spanner API.
+     *
+     * @param array $values The list of values
+     * @return array The encoded values
+     */
+    public function encodeValuesAsSimpleType(array $values)
+    {
+        $res = [];
+        foreach ($values as $value) {
+            $res[] = $this->paramType($value)[0];
+        }
+
+        return $res;
+    }
+
+    /**
+     * Accepts a list of columns (with name and type) and a row from read or
+     * executeSql and decodes each value to its corresponding PHP type.
+     *
+     * @param array $columns The list of columns.
+     * @param array $row The row data.
+     * @return array The decoded row data.
+     */
+    public function decodeValues(array $columns, array $row)
+    {
+        $cols = [];
+        $types = [];
+
+        foreach ($columns as $index => $column) {
+            $cols[] = (isset($column['name']))
+                ? $column['name']
+                : $index;
+            $types[] = $column['type'];
+        }
+
+        $res = [];
+        foreach ($row as $index => $value) {
+            $i = $cols[$index];
```

Comment:
That seems reasonable. I would prefer the default to be RETURN_ASSOCIATIVE since that is the easiest data structure to work with and in most cases people will not have empty or duplicate columns so it should be fine. I would also prefer the method to throw an error if there are duplicate fields and RETURN_ASSOCIATIVE has been provided. Almost always this will be a bug so its better to alert users rather than silently suppress one column. How does RETURN_ASSOCIATIVE handle empty field names?